### DISCUSSION TITLE
[Help]How to implement a parser with the same functionality as this regular expression

### DISCUSSION THREAD

**Discussion by yyy33** [winnow-rs/winnow#559] [Q&A] [answered]
Suppose this is the input text: `"ab hhh fdfv hy tail1 tail2"`, I can use regular expressions(`(.+) (.+) (.+)`) to simply extract the 3 parts I want: `"ab hhh fdfv hy"`, `"tail1"`, `"tail2"`.But it doesn't work for me with the parser:`(take_while(1.., anychar), ' ', (take_while(1.., anychar), ' ', (take_while(1.., anychar))`

**Comment by epage** [comment]
Is that regex literally what needs to be parsed, as in
```rust
let input = "ab hhh fdfv hy tail1 tail2";
let (input, tail2) = input.rsplit_once(" ").unwrap();
let (head, tail1) = input.rsplit_once(" ").unwrap();
```
- `head` has an arbitrary number of spaces in any location, including the front and back
- each tail can be empty or contain just a space

**Reply by yyy33** [reply]
> Is that regex literally what needs to be parsed, as in
> 
> ```rust
> let input = "ab hhh fdfv hy tail1 tail2";
> let (input, tail2) = input.rsplit_once(" ").unwrap();
> let (head, tail1) = input.rsplit_once(" ").unwrap();
> ```
> 
>     * `head` has an arbitrary number of spaces in any location, including the front and back
> 
>     * each tail can be empty or contain just a space

I didn't find rsplit_once as a parser in winnow。

**Reply by epage** [reply]
It isn't.  Those are functions on `&str`.  I was trying to find a way to re-frame the regex to verify its what you actually want.

**Reply by yyy33** [reply]
> It isn't. Those are functions on `&str`. I was trying to find a way to re-frame the regex to verify its what you actually want.

Hi, before mentioning this question, I knew what I wanted, but I had a hard time expressing it, so I gave this similar example to describe my need. Afterwards, I tried to communicate with the AI and finally realized that my need might be:How to implement a regular expression-like backtrace in winnow.

Going back to my example, there are three identical match units in the regular expression(`(.+)`), the first one may eat up the whole input, but since there are still the last two match units, the regular engine assigns reasonable matches to them.

Can we achieve this in winnow:`(parser1, ' ', parser2, ' ', parser3)`,where `parser1 `may eat the entire input.

**Reply by epage** [reply, marked as answer]
The main way people do backtracking is using `alt()` but that requires having distinct fallback cases.  In the situation you gave, there doesn't seem to be distinct fallback cases.  This is why I was asking for clarification to see if that was intended literally or if, with more information, we'd be able to identify fallback cases.

I'm assuming there is more to the need of your parser than just this.  In that case, Winnow is designed to allow you to drop down to writing parse code like this by hand, like the `rsplit_once` code I mentioned.  

One example is
```rust
let (head, tail1, tail2) = preceded('(', take_until(1.., ')')).try_map(|head| {
   let (head, tail2) = head.rsplit_once(' ').ok_or_else(...)?;
   let (head, tail1) = head.rsplit_once(' ').ok_or_else(...)?;
   Ok((head, tail1, tail2))
})?;
```

I would love to find ways to express more power of regexes within winnow and would be open to ideas people have.

**Reply by yyy33** [reply]
> The main way people do backtracking is using `alt()` but that requires having distinct fallback cases. In the situation you gave, there doesn't seem to be distinct fallback cases. This is why I was asking for clarification to see if that was intended literally or if, with more information, we'd be able to identify fallback cases.
> 
> I'm assuming there is more to the need of your parser than just this. In that case, Winnow is designed to allow you to drop down to writing parse code like this by hand, like the `rsplit_once` code I mentioned.
> 
> One example is
> 
> ```rust
> let (head, tail1, tail2) = preceded('(', take_until(1.., ')')).try_map(|head| {
>    let (head, tail2) = head.rsplit_once(' ').ok_or_else(...)?;
>    let (head, tail1) = head.rsplit_once(' ').ok_or_else(...)?;
>    Ok((head, tail1, tail2))
> })?;
> ```
> 
> I would love to find ways to express more power of regexes within winnow and would be open to ideas people have.

Thanks for the help, I'm a newbie and I can only comment from the usage side, as far as I'm concerned I'd like to add a combinator called `backtracking`, like the `alt` combinator, which accepts 1 or more parsers and then automatically backtracks in those parsers to achieve a regular-like effect

**Reply by epage** [reply]
> Thanks for the help, I'm a newbie and I can only comment from the usage side, as far as I'm concerned I'd like to add a combinator called backtracking, like the alt combinator, which accepts 1 or more parsers and then automatically backtracks in those parsers to achieve a regular-like effect

If you have a concrete proposal for how it could do that, I would be open to considering it.

**Reply by yyy33** [reply]
> > Thanks for the help, I'm a newbie and I can only comment from the usage side, as far as I'm concerned I'd like to add a combinator called backtracking, like the alt combinator, which accepts 1 or more parsers and then automatically backtracks in those parsers to achieve a regular-like effect
> 
> If you have a concrete proposal for how it could do that, I would be open to considering it.

Unexpectedly, I am not familiar with the algorithm of regular expressions. But I plan to find time to read the regex crate implementation, and there may be
